### ISSUE TITLE
Major issue: scheduling not working anymore

### ISSUE THREAD

**Issue by JSa1987** [pihome-shc/PiHomeHVAC#214]
Hello,

There is a major issue where the scheduling is not working anymore when there are multiple active zones in the schedule or multiple schedules for the zone. I think I have tracked down this issue to the changes made by @twa127 in #208 .
I believe there currently are two bugs.

**First bug**
I noticed that when a schedule has multiple active zones the same target temperature is used for all zones regardless of what is entered in the schedule for each zone.

I think this due to lines 407-423 in `controller.php`:
```
        // only process active zones
        if ($zone_status == 1) {
                $rval=get_schedule_status($conn, $zone_id,$holidays_status);
                $sch_status = $rval['sch_status'];
                if($rval['sch_count'] == 0){
                        $sch_status = 0;
                        $sch_c = 0;
                        $sch_holidays = '0';
                }else{
                        $sch_end_time = date('H:i:s', $rval['end_time']);
                        $sch_status = $rval['sch_status'];
                        $time_id = $rval['time_id'];
                        $query = "SELECT temperature, coop, holidays_id FROM schedule_daily_time_zone WHERE schedule_daily_time_id = {$time_id} LIMIT 1;";
                        $result = $conn->query($query);
                        $schedule = mysqli_fetch_array($result);
                        $sch_c = $schedule['temperature'];
                        $sch_coop = $schedule['coop'];
```

This pulls the first line from schedule_daily_time_zone for the active schedule, not the line for the zone been processed. I would suggest to change line 419 as follow:

```
        // only process active zones
        if ($zone_status == 1) {
                $rval=get_schedule_status($conn, $zone_id,$holidays_status);
                $sch_status = $rval['sch_status'];
                if($rval['sch_count'] == 0){
                        $sch_status = 0;
                        $sch_c = 0;
                        $sch_holidays = '0';
                }else{
                        $sch_end_time = date('H:i:s', $rval['end_time']);
                        $sch_status = $rval['sch_status'];
                        $time_id = $rval['time_id'];
                        $query = "SELECT temperature, coop, holidays_id FROM schedule_daily_time_zone WHERE schedule_daily_time_id = {$time_id} AND zone_id = {$zone_id} LIMIT 1;";
                        $result = $conn->query($query);
                        $schedule = mysqli_fetch_array($result);
                        $sch_c = $schedule['temperature'];
                        $sch_coop = $schedule['coop'];
```

**Second bug**
When there are multiple schedules defined for a zone in the day the `get_schedule_status()` function in `functions.php` pulls only one. This likely ends up not being the one for the current time and this results in _"Stop Cause: No Schedule"_.
This is caused by `LIMIT 1` on lines 803 and 805 of `functions.php`:

```
function get_schedule_status($conn,$zone_id,$holidays_status){
        // get current day number
        $dow = idate('w');
        // get previous day number, used when end time is less than start time
        $prev_dow = $dow - 1;

        // get raw data
        $query = "SELECT time_id, start, start_sr, Start_ss, Start_offset, end, end_sr, end_ss, end_offset, WeekDays, time_status
                FROM schedule_daily_time_zone_view
                WHERE tz_status = '1' AND zone_id = {$zone_id}";
        if ($holidays_status == 0) {
                $query = $query." AND holidays_id = 0 LIMIT 1;";
        } else {
                $query = $query." AND holidays_id > 0 LIMIT 1;";
        }
```

I would suggest to change line 801 as follow to pull the record only for the current time:

```
function get_schedule_status($conn,$zone_id,$holidays_status){
        // get current day number
        $dow = idate('w');
        // get previous day number, used when end time is less than start time
        $prev_dow = $dow - 1;

        // get raw data
        $query = "SELECT time_id, start, start_sr, Start_ss, Start_offset, end, end_sr, end_ss, end_offset, WeekDays, time_status
                FROM schedule_daily_time_zone_view
                WHERE ((`end`>`start` AND CURTIME() between `start` AND `end`) OR (`end`<`start` AND CURTIME()<`end`) OR (`end`<`start` AND CURTIME()>`start`)) AND zone_id = {$zone_id} AND tz_status = '1' AND (WeekDays & (1 << {$dow})) > 0";
        if ($holidays_status == 0) {
                $query = $query." AND holidays_id = 0 LIMIT 1;";
        } else {
                $query = $query." AND holidays_id > 0 LIMIT 1;";
        }
```

**Comment by twa127** [comment]
thanks I'll incorporate in mysensor fix update

**Comment by JSa1987** [comment]
@twa127 once I fixed the first part of the issue a found a second part and I was editing my original post with the details when you posted the above. I will now give it a try to the fix to the second issue and feedback if it works.

**Comment by JSa1987** [comment]
I have done a little bit of testing and both fixes suggested above seem to be working fine.

**Comment by twa127** [comment]
fixed in #222
